06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+
+### Node State
+When a new reconciliation starts up, a context object is created for each node to store the state and other useful information used by the roller. It will have the following fields:
+
+ - <i>nodeRef</i>: NodeRef object that contains Node ID.
+ - <i>currentNodeRole</i>: Currently assigned process roles for this node (e.g. controller, broker).
+ - <i>lastKnownState</i>: It contains the last known state of the node based on information collected from the abstracted sources (Kubernetes API, KafkaAgent and Kafka Admin API). The table below describes the possible states.
+ - <i>restartReason</i>: It is updated based on the current predicate logic from the `Reconciler`. For example, an update in the Kafka CR is detected.
+ - <i>numRestartAttempts</i>: The value is incremented each time the node has been attempted to restart.
+ - <i>numReconfigAttempts</i>: The value is incremented each time the node has been attempted to reconfigure.
+ - <i>numRetries</i>: The value is incremented each time the node cannot be restarted/reconfigured due to not meeting safety conditions (more on this later).
+ - <i>lastTransitionTime</i>: System.nanoTime of last observed state transition.
+
+ The following table illustrates the proposed node states and the possible transitions:
+ | State            | Description      | Possible transitions |
+ | :--------------- | :--------------- | :----------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running). This is determined via Kubernetes API, more details for it below. | `NOT_READY`, `NOT_RUNNING`. | `RESTARTED` `SERVING` | 
+ | NOT_READY             | Node is running but not ready to serve requests which is determined by Kubernetes readiness probe (broker state < 2 OR == 127 OR controller is not listening on port). | `RESTARTED` `SERVING` |
+ | RESTARTED             | After successful `kubectl delete pod`. | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | RECONFIGURED          | After successful Kafka node config update via Admin client. | `NOT_RUNNING` `NOT_READY` `RESTARTED` `SERVING` |
+ | RECOVERING            | Node has started but is in log recovery (broker state == 2). | `SERVING` |
+ | SERVING               | Node is in running state and ready to serve requests which is determined by Kubernetes readiness probe (broker state >= 3 AND != 127 OR controller is listening on port). | `RESTARTED` `RECONFIGURED` `LEADING_ALL_PREFERRED` |
+ | LEADING_ALL_PREFERRED | Node is in running state and leading all preferred replicas. | None
+
+The definitions of broker states can be found via the following link: [Broker States](https://github.com/apache/kafka/blob/3.7/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java).
```

Comment:
But currently we do rely on the broker states, to do the readiness check and find if node is doing log recovery. I can add that this may change or not mention broker states at all.

Reply:
Basically, I do not understand why the link is here:
* Are you just saying it was used as an source to know what states there are?
* Does it mean we need to update the code anytime the state changes to add our own new states?
* Will we include this class directly in our code and use the states from it?

Reply:
> Are you just saying it was used as an source to know what states there are?

Yes, just to give a context when we talk about broker states in general.

> Does it mean we need to update the code anytime the state changes to add our own new states?

Unlikely. Unless enums of the broker states change upstream and suddenly they mean different things. As long as the change don't affect the readiness check and recovery state, we wouldn't need to worry about it I think. 

> Will we include this class directly in our code and use the states from it?

No. The link is there just there as an information. How we check the broker state is not changing in the new roller. It's same as the current roller (via KafkaAgent).

Reply:
I can clarify why the link is shared or remove the link.